metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/impl/EntityClientResolver.java

```diff
@@ -0,0 +1,174 @@
+
+    private EntityClientApi entityClientApi;
+    private final int batchSize;
+
+
+    public EntityClientResolver(int batchSize) {
+        this.batchSize = batchSize;
+        entityClientApi = new EntityClientApiImpl();
+    }
+
+    @Override
+    public <T extends SearchTerm> Map<T, List<EnrichmentBase>> resolveByText(Set<T> searchTerms) {
+        final Function<T, EnrichmentQuery> inputFunction = EntityResolverUtils.createEnrichmentQueryForTextSearch();
+        return performInBatches(searchTerms, inputFunction);
+    }
+
+    @Override
+    public <T extends ReferenceTerm> Map<T, EnrichmentBase> resolveById(Set<T> referenceTerms) {
+        final Function<T, EnrichmentQuery> inputFunction = EntityResolverUtils.createEnrichmentQueryForRefSearch();
+        return EntityResolverUtils.pickFirstFromTheList(performInBatches(referenceTerms, inputFunction));
+    }
+
+    @Override
+    public <T extends ReferenceTerm> Map<T, List<EnrichmentBase>> resolveByUri(Set<T> referenceTerms) {
+        final Function<T, EnrichmentQuery> inputFunction = EntityResolverUtils.createEnrichmentQueryForRefSearch();
+        return performInBatches(referenceTerms, inputFunction);
+    }
+
+    /**
+     * Gets the Enrichment for Multiple input values in batches of batchSize
+     *
+     * @param inputValues Set of values for which enrichment will be performed
+     * @param requestParser Function to return EntityClientRequest from input value <I>
+     * @param <I> Input value Type. Could be <T extends SearchTerm> OR <T extends ReferenceTerm>
+     * @return
+     */
+    private <I> Map<I, List<EnrichmentBase>> performInBatches(Set<I> inputValues, Function<I, EnrichmentQuery> requestParser) {
+        final Map<I, List<EnrichmentBase>> result = new HashMap<>();
+        // create partitions
+        final List<List<I>> partitions = EntityResolverUtils.createBatch(inputValues, batchSize);
+
+        // Process partitions
+        for (List<I> partition : partitions) {
+            for (I input : partition) {
+                EnrichmentQuery clientRequest = requestParser.apply(input);
+                List<EnrichmentBase> enrichmentBaseList = executeEntityClientRequest(clientRequest);
+                if (!enrichmentBaseList.isEmpty()) {
+                    result.put(input, enrichmentBaseList);
+                }
+            }
+        }
+        return result;
+    }
+
+    private List<EnrichmentBase> executeEntityClientRequest(EnrichmentQuery clientRequest) {
+        // 1. get Entities
+        List<Entity> entities = getEntities(clientRequest);
+        if (!entities.isEmpty()) {
+            // 2. get the parent entities
+            List<Entity> parentEntities = new ArrayList<>();
+            entities.stream().forEach(entity -> {
+                if (entity.getIsPartOfArray() != null && EntityResolverUtils.isTextOrUriSearch(clientRequest)) {
+                    fetchParentEntities(entity, parentEntities);
```

Comment:
Here is the ticket for Parent entity logic - [EA-2984](https://europeana.atlassian.net/browse/EA-2984). It is mentioned to fetch parent entities for both text and Uri search. One thing I am getting confused about is the service side and client side. Are the functionality of enrichment different in service and client ??

Reply:
Ok. But should this be done here or inside the service?
I'm asking because for each items the parents have to be resolved remotely while it could be done much faster inside the service.
Also how about co-referencing logic??

Reply:
Well, for us performance is important. So if we will gain better results doing this remotely, we should do this remotely in the service. 
For co-referencing logic, probably best to ask Hugo this.  From the ticket, I only know that we needed to fetch the parents for the entities.

Reply:
I resolved the other similar comment.
So we need some decisions on the following:
- fetch parents implementation location. Now it's remotely performed, but could be in the Entity API application.
- Is the timespam broadness check implemented in the Entity API application? If not, should it? (metis checks that)
- Is the co-referencing implemented for URI searches in the Entity API application? If not, should it? (metis checks that)

Reply:
After discussion:
- Fetching parents should be applied for text and uri searches but not id searches to match the current metis-enrichment implementation. (Id parent fetching can be discussed/implemented later)
And the implementation should, for now, stay in the metis-enrichment code(if the re-processing operation shows a big decline on speed, we should then try to move the parent fetching on the Entity API application).
- Timespan broadness check is not required.
- For URI searches, the about check and then owlSameAs check is implemented with the reference.startsWith(EntityApiConstants.BASE_URL) where if it matches  it will use the entityClientApi.getEntityById(referenceValue) otherwise the entityClientApi.getEntityByUri(referenceValue);

Reply:
I've added the parent searching for all search cases (id, text, uri)